### PR TITLE
[Instrumentation.Http] Support for enrich and filter delegates for http client metric

### PR DESCRIPTION
**Pull request by aunikitin**
Hello, I'd like to have an opportunity to enrich metrics with custom logic (for example uri templates of requested paths). Also I add opportunity to filter some requests because of **TODO** in `MeterProviderBuilderExtensions`
I tried to use existing `HttpClientInstrumentationOptions` but firstly, they have too mutch for metrics and secondly, because in the `HttpHandlerMetricsDiagnosticListener` we are not allowed to take tags from activity (because of uncontrolled source of filling), that's why delegates in settings don't fit .
Also I wrote some tests and did small refactoring in them (move trace tests in one file; split configuration tests from common, because they don't need running `HttpListener`).
If my changes are admissible, I will add description to `CHANGELOG.md`.

Probably it resolves [issue](https://github.com/open-telemetry/opentelemetry-dotnet/issues/3441)

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by aunikitin** [comment]
This is my first deal with EasyCLA and unfortunately, I don't know what to do :(  
I have already signed form by link from bot comment, but it has no effect

**Review by vishweshbankwar — CHANGES_REQUESTED** [review]
@aunikitin Thanks for your PR. Could you please update the PR to only include changes related to enrich and filter options on metrics?. This will keep it small and easier to review. Any refactoring, renaming can be done in a separately.

**Comment by aunikitin** [comment]
@vishweshbankwar hello, you want me to rollback changes in tests (for example moving test `AddHttpClientInstrumentationUsesOptionsApi` to `ProviderBuilderExtensionsTests` withour `HttpListener` and etc.) or only options renaming?

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs [hunk 1]
```diff
@@ -29,41 +29,108 @@ internal sealed class HttpHandlerMetricsDiagnosticListener : ListenerHandler
...
+                return;
+            }
+
+            var request = response.RequestMessage;
+            if (!this.TryFilterHttpRequestMessage(activity.OperationName, request))
+            {
+                return;
+            }
+
+            var tags = new List<KeyValuePair<string, object>>
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:68]: we should try to minimize heap allocations in hot path.

Taking a step back, we do not have any way to achieve "Enrich" equivalent in the SDK itself today, so this needs to be first design reviewed before making it part of the instrumentation library.

**Review by cijothomas — CHANGES_REQUESTED** [review]
See https://github.com/open-telemetry/opentelemetry-dotnet/pull/4075/files#r1073154671

Not yet ready to make new additions to instrumentation library as we do not have a metrics enrichment concept yet.

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:68]: Spoke to @vishweshbankwar offline about recent changes. We can do this for now (as this was already done for other instrumentation). Please make sure the allocation is not done by following similar approach as other instrumentation (aspnetcore).

Comment (aunikitin) [reply src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs]: Hello, nice idea, I missed opportunity to create custom delegate while doing

**Comment by cijothomas** [comment]
@aunikitin Thanks for the PR. I have triggered CI checks and will review. 
There is some pending discussions about potentially making stable releases of Instrumentation Libraries, and we'll need to evaluate all the public APIs at that time, and the Enrich/Filter will be re-evaluated at that time to make sure its in alignment with overall Metrics design. (See https://github.com/open-telemetry/opentelemetry-dotnet/pull/3948#issuecomment-1334144563 as well)

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    /// <param name="name">The name of the metric being enriched.</param>
+    /// <param name="requestMessage"><see cref="HttpRequestMessage"/>: the HttpRequestMessage object.</param>
+    /// <param name="tags"><see cref="TagList"/>: List of current tags. You can add additional tags to this list. </param>
+    public delegate void HttpRequestMessageEnrichmentFunc(string name, HttpRequestMessage requestMessage, ref TagList tags);
+
+    /// <summary>
+    /// Gets or sets an action to enrich an tags collection with <see cref="HttpRequestMessage"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para><b>EnrichWithHttpRequestMessage is only executed on .NET and .NET
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs:42]: what is this comment ? its looks like a copy paste error from other library 🤣

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs:42]: oh my bad. read it fully now only.

**Comment by aunikitin** [comment]
@cijothomas okay, I'm waiting of your decision.
Can you restart CI? I faced with some issues under Windows, hope they're fixed

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs [hunk 3]
```diff
@@ -26,44 +26,110 @@ namespace OpenTelemetry.Instrumentation.Http.Implementation
...
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpFlavor, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version)));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerName, request.RequestUri.Host));
+            if (!this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) || response == null)
+            {
+                // TODO: logging?
+                return;
+            }
+
+            var request = response.RequestMessage;
+            if (!this.TryFilterHttpRequestMessage(activity.OperationName, request))
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:64]: this looks like response message and not request?

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs [hunk 4]
```diff
@@ -26,44 +26,110 @@ namespace OpenTelemetry.Instrumentation.Http.Implementation
...
+            if (!request.RequestUri.IsDefaultPort)
+            {
+                tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerPort, request.RequestUri.Port));
+            }
 
-                    if (!request.RequestUri.IsDefaultPort)
-                    {
-                        tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerPort, request.RequestUri.Port));
-                    }
+            this.EnrichWithHttpRequestMessage(ref tags, request);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:81]: same as above...this looks like response message and not request?

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Diagnostics;
+using System.Net;
+#if NETFRAMEWORK
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs:19]: where is this used?

File: src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs [hunk 6]
```diff
@@ -30,21 +32,52 @@ public static class MeterProviderBuilderExtensions
...
+            {
+                builder.ConfigureServices(services => services.Configure(optionsName, configureOptions));
+            }
 
-            // TODO: Handle HttpClientInstrumentationOptions
-            //   SetHttpFlavor - seems like this would be handled by views
-            //   Filter - makes sense for metric instrumentation
-            //   Enrich - do we want a similar kind of functionality for metrics?
-            //   RecordException - probably doesn't make sense for metric instrumentation
+            builder.ConfigureBuilder((sp, builder) =>
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:72]: Could we keep this to a separate change, as this relies to newest SDK features, and should not be a pre-req for this PR?

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+/// </summary>
+public class HttpClientInstrumentationMeterOptions
+{
+    /// <summary>
+    /// Delegate for enrichment of recorded metric with additional tags.
+    /// </summary>
+    /// <param name="name">The name of the metric being enriched.</param>
+    /// <param name="requestMessage"><see cref="HttpRequestMessage"/>: the HttpRequestMessage object.</param>
+    /// <param name="tags"><see cref="TagList"/>: List of current tags. You can add additional tags to this list. </param>
+    public delegate void HttpRequestMessageEnrichmentFunc(string name, HttpRequestMessage requestMessage, ref TagList tags);
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs:36]: I think we need to pass in both request and response.
https://github.com/open-telemetry/opentelemetry-dotnet/pull/4098
@cijothomas - FYI

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs (hunk 4 above)
Comment (aunikitin) [reply src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:81]: In meter, both request and response are used. Should I do the same? I agree that it's more convinient to have response, but it'll conflict with https://github.com/open-telemetry/opentelemetry-dotnet/pull/4098

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs (hunk 5 above)
Comment (aunikitin) [reply src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs]: remove this, I left only `System.Net.Http` dependency for `HttpResponseMessage`

File: src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs (hunk 6 above)
Comment (aunikitin) [reply src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:72]: as you wish

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs (hunk 7 above)
Comment (aunikitin) [reply src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationMeterOptions.cs]: I'll leave HttpResponse, if your changes appear earlier, I'll fix mine, and vice versa

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs (hunk 3 above)
Comment (aunikitin) [reply src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs]: replace with response

**Pull request closed without merging**